src/main/java/io/confluent/connect/jdbc/dialect/GenericDatabaseDialect.java

```diff
@@ -0,0 +1,1727 @@
+  private final IdentifierRules defaultIdentifierRules;
+  private final AtomicReference<IdentifierRules> identifierRules = new AtomicReference<>();
+  private final Queue<Connection> connections = new ConcurrentLinkedQueue<>();
+  private volatile JdbcDriverInfo jdbcDriverInfo;
+
+  /**
+   * Create a new dialect instance with the given connector configuration.
+   *
+   * @param config the connector configuration; may not be null
+   */
+  public GenericDatabaseDialect(AbstractConfig config) {
+    this(config, IdentifierRules.DEFAULT);
+  }
+
+  /**
+   * Create a new dialect instance with the given connector configuration.
+   *
+   * @param config                 the connector configuration; may not be null
+   * @param defaultIdentifierRules the default rules for identifiers; may be null if the rules are
+   *                               to be determined from the database metadata
+   */
+  protected GenericDatabaseDialect(
+      AbstractConfig config,
+      IdentifierRules defaultIdentifierRules
+  ) {
+    this.config = config;
+    this.defaultIdentifierRules = defaultIdentifierRules;
+    this.jdbcUrl = config.getString(JdbcSourceConnectorConfig.CONNECTION_URL_CONFIG);
+    if (config instanceof JdbcSinkConfig) {
+      catalogPattern = JdbcSourceTaskConfig.CATALOG_PATTERN_DEFAULT;
+      schemaPattern = JdbcSourceTaskConfig.SCHEMA_PATTERN_DEFAULT;
+      tableTypes = new HashSet<>(Arrays.asList(JdbcSourceTaskConfig.TABLE_TYPE_DEFAULT));
+    } else {
+      catalogPattern = config.getString(JdbcSourceTaskConfig.CATALOG_PATTERN_CONFIG);
+      schemaPattern = config.getString(JdbcSourceTaskConfig.SCHEMA_PATTERN_CONFIG);
+      tableTypes = new HashSet<>(config.getList(JdbcSourceTaskConfig.TABLE_TYPE_CONFIG));
+    }
+    if (config instanceof JdbcSourceConnectorConfig) {
+      mapNumerics = ((JdbcSourceConnectorConfig)config).numericMapping();
+    } else {
+      mapNumerics = NumericMapping.NONE;
+    }
+  }
+
+  @Override
+  public Connection getConnection() throws SQLException {
+    // These config names are the same for both source and sink configs ...
+    String username = config.getString(JdbcSourceConnectorConfig.CONNECTION_USER_CONFIG);
+    Password dbPassword = config.getPassword(JdbcSourceConnectorConfig.CONNECTION_PASSWORD_CONFIG);
+    Properties properties = new Properties();
+    if (username != null) {
+      properties.setProperty("user", username);
+    }
+    if (dbPassword != null) {
+      properties.setProperty("password", dbPassword.value());
+    }
+    properties = addConnectionProperties(properties);
+    Connection connection = DriverManager.getConnection(jdbcUrl, properties);
+    if (jdbcDriverInfo == null) {
+      jdbcDriverInfo = createJdbcDriverInfo(connection);
+    }
+    connections.add(connection);
+    return connection;
+  }
+
+  @Override
+  public void close() {
+    for (Connection conn = connections.poll(); conn != null; conn = connections.poll()) {
+      try {
+        conn.close();
```

Comment:
Maybe use `org.apache.kafka.common.utils.Utils.closeQuietly`, which will catch `Throwable`?  It seems if `close()` throws a `Throwable`, the other connections will not get cleaned up.  This could cause issues on the server side.